diag/diagnostics_test.go

```diff
@@ -0,0 +1,527 @@
+			t.Parallel()
+
+			tc.diags.AddWarning(tc.summary, tc.detail)
+
+			if diff := cmp.Diff(tc.diags, tc.expected); diff != "" {
+				t.Errorf("Unexpected response (+wanted, -got): %s", diff)
+			}
+		})
+	}
+}
+
+func TestDiagnosticsAppend(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		diags    diag.Diagnostics
+		in       diag.Diagnostics
+		expected diag.Diagnostics
+	}{
+		"nil-append": {
+			diags: nil,
+			in: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+			},
+			expected: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+			},
+		},
+		"append": {
+			diags: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+			},
+			in: diag.Diagnostics{
+				diag.NewErrorDiagnostic("three summary", "three detail"),
+				diag.NewWarningDiagnostic("four summary", "four detail"),
+			},
+			expected: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+				diag.NewErrorDiagnostic("three summary", "three detail"),
+				diag.NewWarningDiagnostic("four summary", "four detail"),
+			},
+		},
+		"empty-diagnostics": {
+			diags: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+			},
+			in: nil,
+			expected: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+			},
+		},
+		"empty-diagnostics-elements": {
+			diags: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+			},
+			in: diag.Diagnostics{
+				nil,
+				nil,
+			},
+			expected: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+			},
+		},
+		"duplicate": {
+			diags: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+			},
+			in: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+			},
+			expected: diag.Diagnostics{
+				diag.NewErrorDiagnostic("one summary", "one detail"),
+				diag.NewWarningDiagnostic("two summary", "two detail"),
+			},
+		},
+	}
+
+	for name, tc := range testCases {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			tc.diags.Append(tc.in...)
+
+			if diff := cmp.Diff(tc.diags, tc.expected); diff != "" {
+				t.Errorf("Unexpected response (+wanted, -got): %s", diff)
+			}
+		})
+	}
+}
+
+func TestDiagnosticsContains(t *testing.T) {
```

Comment:
Needs some test cases with attribute paths

Reply:
Added in https://github.com/hashicorp/terraform-plugin-framework/pull/110/commits/473b6e1ac4ac0c34a2ef9faa033bf627b4f2f041 -- please let me know if you want anything else!